src/lowered.jl

```diff
@@ -0,0 +1,173 @@
+## Analyzing lowered code
+
+function add_docexpr!(docexprs::AbstractDict{Module,V}, mod::Module, ex) where V
+    docexs = get(docexprs, mod, nothing)
+    if docexs === nothing
+        docexs = docexprs[mod] = V()
+    end
+    push!(docexs, ex)
+    return docexprs
+end
+
+function lookup_callexpr(frame, stmt)
+    fargs = JuliaInterpreter.collect_args(frame, stmt)
+    return Expr(:call, fargs...)
+end
+
+function assign_this!(frame, value)
+    frame.framedata.ssavalues[frame.pc] = value
+end
+
+# This defines the API needed to store signatures using methods_by_execution!
+# This default version is simple; there's a more involved one in Revise.jl that interacts
+# with CodeTracking.
+const MethodInfo = IdDict{Type,LineNumberNode}
+add_signature!(methodinfo::MethodInfo, sig, ln) = push!(methodinfo, sig=>ln)
+push_expr!(methodinfo::MethodInfo, mod::Module, ex::Expr) = methodinfo
+pop_expr!(methodinfo::MethodInfo) = methodinfo
+
+function methods_by_execution(mod::Module, ex::Expr; define=true)
+    methodinfo = MethodInfo()
+    docexprs = Dict{Module,Vector{Expr}}()
+    value = methods_by_execution!(finish_and_return!, methodinfo, docexprs, mod, ex; define=define)
+    return methodinfo, docexprs
+end
+
+function methods_by_execution!(@nospecialize(recurse), methodinfo, docexprs, mod::Module, ex::Expr; define=true)
+    try
+        frame = prepare_thunk(mod, ex)
+        frame === nothing && return nothing
+        return methods_by_execution!(recurse, methodinfo, docexprs, frame; define=define)
+    catch err
+        isa(err, InterruptException) && rethrow(err)
+        @warn "error evaluating in module $mod: $ex"
```

Comment:
```suggestion
        @warn "error evaluating in module $mod" exception=ex
```

Loggers implement special handling/rendering for the `exception` kwarg.
Can probably also remove the `showerror` below, because of this.